[WIP] Convert Exporter Customization Dialog to javafx

NorwayMaple opened the pull request:
This is my work so far.  I am just beginning the dialog conversion and wanted to share my reasoning so far.  What I am doing is listed in the comments in class `ExportCustomizationDialogViewModel`.

<!-- describe the changes you have made here: what, why, ... 
     Link issues by using the following pattern: [#333](https://github.com/JabRef/jabref/issues/333) or [koppor#49](https://github.com/koppor/jabref/issues/47).
     The title of the PR must not reference an issue, because GitHub does not support autolinking there. -->


----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Manually tested changed features in running JabRef
- [ ] Screenshots added in PR description (for bigger UI changes)
- [ ] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

Siedlerchr commented on src/main/java/org/jabref/preferences/PreferencesService.java:
```diff
@@ -35,4 +35,8 @@
 
     public void updateEntryEditorTabList();
 
+    List<String> getStringList(String key);
```
Please no generic methods here, create specific ones for the cases you need, e.g. get CustomExporters or getExporterormat...  Or you create CustomExporter preferences which has all the needed information inside.

NorwayMaple replied:
The Swing version has a class CustomExportList which holds a list of custom exporters and associated data.  I was going to move some of the code here into the ExporterViewModel and other code into the ExporterCustomizationDialogViewModel.  Do you recommend keeping the CustomExportList as it is?

tobiasdiez replied:
The PreferencesService should only contain usecase-specific preference methods and the caller shouldn't care about how these information are stored in the preferences. So, here, the method would look like `List<Exporter> getCustomExportFormats()` and then in the implementation there would be calls like `preferences.getStringList(JabRefPreferences.CUSTOM_EXPORT_FORMAT + i))`.

Thus, most of the code in `CustomExportList` should go into `JabrefPreferences`.

NorwayMaple replied:
Shouldn't the methods in `preferencesService` match those invoked in `ExportCustomizationDialogViewModel` as `exporters.add(preferences.X)` where X is the method, rather than using `preferences.getStringList`?

tobiasdiez replied:
Sorry, I don't really understand what you mean. The PreferencesService should provide an interface to the preferences that hide the exact semantics of how the object is serialized. So abstractly, `X getX()` and `putX(X newValue)`, where `X` is the (model) object. The implementation in `JabRefPreferences` than has to worry about (de)serialization.

tobiasdiez replied:
I'm not a big fan of writing tests for the sake of it. My strategy is usually to program everything and when I discover a problem in the logic during testing, I'll write a test for it. Hence, if the dialog is simple enough that you don't make any errors, then you don't need to write tests ;-).
An exception is when there is really complex logic that is definitely worth testing.

tobiasdiez commented on src/main/java/org/jabref/gui/exporter/ExportCustomizationDialogViewModel.java:
```diff
@@ -0,0 +1,71 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ExportCustomizationDialogViewModel extends BaseDialog<Void> {
```
The `View`/Dialog class should derive from `BaseDialog` and not the view model.

tobiasdiez commented on src/main/java/org/jabref/gui/exporter/ExportCustomizationDialogViewModel.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    //possibly not necessary, and if so getSortedList will have to return the correct type, not Eventlist<List<String>>
+    public List<ExporterViewModel> loadExporters() {
```
You might don't need the gui-wrapper `ExporterViewModel` around `Exporter`. In general such a wrapper is needed to adapt the properties of the model to display it to the user. But, here, the data of an exporter is relatively simple and the already in a suitable format for the gui.

NorwayMaple replied:
I have put the data in the GUI wrapper `ExporterViewModel` for now.  How would I create the `TableView` and `TableColumns` without a `ViewModel` for the data?  All the examples I could find in JabRef and online had some sort of `ViewModel`.

tobiasdiez replied:
You can also simply use `Exporter` as the datatype for the TableView. But if you have already created a ViewModel for it, keep it. It is probably the cleaner approach and more versatile for the future.

NorwayMaple replied:
Okay.  I've finished, but not tested, most of the `ViewModel`.  I just have to add the dialog to add or modify the exporter, and then create the View and FXML.  Unless there's something I forgot...

NorwayMaple commented:
At this point, I've refactored most of the code from the dialogs.  The reasoning is mostly implemented already or in the comments.  Take a look if you have time and let me know what you think!  Otherwise, I can keep going and let you know when I'm done.

Siedlerchr commented on src/main/java/org/jabref/gui/exporter/CreateModifyExporterDialogViewModel.java:
```diff
@@ -0,0 +1,89 @@
+            .addExtensionFilter(Localization.lang("Custom layout file"), StandardFileType.LAYOUT)
+            .withDefaultExtension(Localization.lang("Custom layout file"), StandardFileType.LAYOUT)
+            .withInitialDirectory(getExportWorkingDirectory()).build();
+        dialogService.showFileOpenDialog(fileDialogConfiguration).ifPresent(f -> setText(f.toAbsolutePath().toString())); //implement setting the text
```
To set the text, you would create a StringProperty here, e.g. selectedPath and then bind that property to a textfield in the View

NorwayMaple replied:
Fixed it.  I think I had the property, but I hadn't realized this is what I needed here.  Thanks for your recommendation!

Siedlerchr commented on src/main/java/org/jabref/gui/exporter/ExportCustomizationDialogViewModel.java:
```diff
@@ -0,0 +1,91 @@
+
+    }
+
+    public void modifyExporter(int row) {
```
Do you really need an int here? It would be generally better if you just operate on the row object

NorwayMaple replied:
I don't think I need it.  I changed it to use a property `selectedExporters`.

NorwayMaple commented:
Hi, I am having trouble using EasyBind to bind a `ListProperty` in the view model to a table view in the View.  Specifically, I want to bind `viewModel.selectedExportersProperty` to the selected rows in `exporterTable` within class `ExportCustomizationDialogView`.  I am using `KeyBindingsDialogView` as an example of how to use EasyBind.  However, `KeyBindingsDialogView` uses a `treeTable` structure in lines 56-59, which is not what I want in my dialog.  I have spent at least four hours today reading about JavaFX, data binding, properties, observers, and other things, but I still can't figure it out.  I have some ideas on the recent pull request, but Eclipse still throws compiler errors.  Any ideas on what I should do?

Siedlerchr commented:
@NorwayMaple  No need to complex binding.  Just use setItems.
I created a simple patch fixing some of the issues you mentioned and generated some stubs.
Remove the .txt extensions and import it with `git am < file.patch`

Another thing I just noticed, for the SimpleXYZProperty in javafx, there exists always an interface, e.g. XYZ Property. Always use that.

[0001-fix-some-binding-issues.patch.txt](https://github.com/JabRef/jabref/files/2544172/0001-fix-some-binding-issues.patch.txt)

NorwayMaple commented:
Thank you!  I took a brief look at it, and it looks good.  In the future, how can I decide when to use EasyBind?

NorwayMaple commented:
I finished the View and am trying to load it from the JabRef menu, but Afterburner throws an error:
```
14:44:21.187 [JavaFX Application Thread] ERROR org.jabref.FallbackExceptionHandler - Uncaught exception occurred in Thread[JavaFX Application Thread,5,main] - java.lang.IllegalStateException: Cannot load org.jabref.gui.exporter.exportcustomizationdialog
	at com.airhacks.afterburner.views.ViewLoader.load(ViewLoader.java:174)
```
etc...

I grepped `exportcustomizationdialog` and `ExportCustomizationDialog`, which is the old Swing version, in `src/`, and all references to the old Swing dialog are in the Swing dialog itself.  Any ideas what is going on?

NorwayMaple commented:
Thanks for the help!  Those corrections both make sense.  I was not seeing that it was referring to the FXML.  I made some corrections.  However, it still won't load.

Also, how do you pronounce JabRef?  I tell my friends I'm working on it, and sometimes I say "Jab-
Ref," and other times "Job-ref."

NorwayMaple commented:
I've got it to load (although there is work to do).  I'll let you know if I have more questions (besides the pronunciation)!

Siedlerchr commented:
Glad that it works! 😎👍Don't hesitate to ask here or in gitter. We try to help then. 

I pronounce it as Jab Ref. We Germans tend so speak as it is written. Actually the history of the name is quite unique. 

JabRef started in 2003 after the authors of BibKeeper (Morton Alver) and JBibtexManager (Nizar Batada) merged their projects.

And the name is J for Java, A for Alver and B for Batada and Ref for Reference.

A bit lesser known Trivia.

NorwayMaple commented:
Interesting!  So "a" is not like "a" in "Guten tag"?

Siedlerchr commented:
@NorwayMaple  Depends on the word, vocals can be either short or long.  I think my pronunciation is more like Dschapp-Ref (literally). But in the end it doesn't matter. Pronounce it like you want as long as it works ;)

NorwayMaple commented:
I have two questions:

1.
I read pull request #3092 and it seemed to say we are trying to move away from using `JabRefPreferences` and toward using separate preferences objects like `LayoutPreferences` and `SavePreferences`, that can be passed into logic and model objects.  Should I be using different preferences objects as well rather than `JabRefPreferences`?

2.
I am trying to move code out of `CustomExportList` so that I can get rid of the class.  As of now I have taken lines 91-102 in that class and largely put them (with modifications) into `JabRefPreferences.getCustomExportFormats` lines 2034-2046 and `CreateModifyExporterDialogViewModel` (my VM for the new subdialog for custom exporters) lines 80-91.  I also need similar functionality in `ExporterViewModel` around line 17.  To avoid the boilerplate processing of strings and filetypes in all three classes before instantiation of `TemplateExporter`, which is in logic, I was thinking of putting a new constructor into that class (which I started doing).  Is this recommended?

I can also spend more time reading methods that call `CustomExportList` and methods in `TemplateExporter` tomorrow if you're not sure.

Thanks!

Siedlerchr commented:
There is no general rule of thumb here, I think if you have more than one preference for a dialog or whatever, e.g. you have multiple preferences you acesss, as with the Timestamp prefs, it is useful to have them grouped into one single Preferences Class.

PS: You should merge master in and resolve the conflicts, to stay up to date

NorwayMaple commented:
@tobiasdiez @Siedlerchr 

Hi guys,

I just finished a minimal version of the dialog that works.  It should be able to add, modify, remove, and save custom exporters to preferences, as well as load from preferences.

The major architectural change was that I removed `org.jabref.preferences.CustomExportList` and put its functionality into `JabRefPreferences` and logic because I found the original class to be incompatible with MVVM architecture.  The changes to logic were minimal and included changes to `TemplateExporter` and `StandardFileType` that allow more flexible inputs.  Through this, I have simplified saving and loading from preferences.

Take a look and/or test it, and let me know what you think.  I can improve it by adding invalid input checks, etc., but as long as the user plays along, it works so far.  I also can make the list sorted alphabetically, but I haven’t yet.

You may have to click the save button a few times for it to work - I’m not sure why.

Siedlerchr commented on src/main/java/org/jabref/gui/exporter/CreateModifyExporterDialogViewModel.java:
```diff
@@ -0,0 +1,120 @@
+
+        setTextFields(exporter);
+
+
```
I am sure checkstyle will complain about too many empty lines(not more than one)

NorwayMaple replied:
I am fixing spacing and putting it in my next push.  Thanks!

Siedlerchr commented on src/main/java/org/jabref/preferences/JabRefPreferences.java:
```diff
@@ -1983,4 +1983,70 @@ public FileDragDropPreferenceType getEntryEditorFileLinkPreference() {
+        List<String> s;
+        // possibly check if CUSTOM_EXPORT_FORMAT + 0 is empty too and throw error
+        while (!((s = getStringList(CUSTOM_EXPORT_FORMAT + i)).isEmpty())) {
+            exporterName = s.get(0);
```
Pleas use constants for the indicies, e.g.  EXPORTER_NAME_INDEX = 0
And please use a better name for "s"

NorwayMaple replied:
I have fixed these issues.

Siedlerchr commented on src/main/java/org/jabref/logic/util/StandardFileType.java:
```diff
@@ -52,6 +52,11 @@
     }
 
     public static FileType newFileType(String... extensions) {
+        for (int i = 0; i < extensions.length; i++) {
```
Could you please explain to me why this is necessary?
You remove the dot from extension?

NorwayMaple replied:
No problem!  Since I am new, if you have another idea of how to implement this, let me know.

If you look in lines 97-100 in `org.jabref.preferences.CustomExportList` on the master branch, you will see that there is a check for the `.` while processing extensions passed into the class, usually as object `Globals.prefs.customImports` (also removed in PR branch).  In the `master` branch, this is because the TemplateExporter constructor takes a `FileType` without a `.`  Since this is text processing, and could be relevant to any class that wants to use `StandardFileType`, I decided to put it into logic.  This way, that check is done inside the `StandardFileType`.  It doesn't impose any new requirements, so it should not break anything else.  That said, I created a new constructor in `TemplateExporter` and could put it there instead.

As a side note, I forgot to reference the new constructor (with fewer parameters), and have pushed changes referencing it.

NorwayMaple commented:
As for those checks, they were already implemented, in that the dialog just returns an empty `Optional` in those cases, among others.  The code is as follows:

```
        if (layoutFile.get().isEmpty() || name.get().isEmpty() || extension.get().isEmpty()
            || !layoutFile.get().endsWith(".layout")) {

            LOGGER.info("One of the fields is empty or invalid!");
            // Return empty exporter to the main exporter customization dialog
            return Optional.empty();
        }
```

NorwayMaple commented:
Sorry, I closed the PR by mistake, so I reopened it.

NorwayMaple commented:
@Siedlerchr @tobiasdiez If it would be helpful, I can go through the PR briefly on Skype.  That way I could go through the code with you step-by-step.

Siedlerchr commented:
@NorwayMaple  Sure, you can add me in skype under siedlerkiller2.0 then we can make up a time for a call.

Siedlerchr commented on src/main/java/org/jabref/gui/exporter/ExportCustomizationDialogViewModel.java:
```diff
@@ -0,0 +1,98 @@
+public class ExportCustomizationDialogViewModel extends AbstractViewModel {
+
+    //TODO: Use JavaFX SortedList instead of SimpleListPreperty for exporters.  There is no SimpleSortedListProperty.
+    private final ListProperty<ExporterViewModel> exporters = new SimpleListProperty<>(FXCollections.observableArrayList());
```
A sorted list is read only, you need to wrap it around an existing observable list

Siedlerchr commented on src/main/java/org/jabref/gui/exporter/ExporterViewModel.java:
```diff
@@ -0,0 +1,49 @@
+        this.lfFileName.setValue(exporter.getLayoutFileNameWithExtension());
+        // Only the first of the extensions gotten from FileType is saved into the class using get(0)
+        // substring(1) used to remove "*" from extension, i.e. "*.txt" to ".txt"
+        String extensionString = exporter.getFileType().getExtensionsWithDot().get(0).substring(1);
```
I would move this method as getFirstExtensionWithoutDot into FileType

NorwayMaple replied:
Actually, it turns out that there was a method `getExtensions()` in `FileType` that does the job.  Pushed.

Siedlerchr commented on src/main/java/org/jabref/logic/util/StandardFileType.java:
```diff
@@ -52,6 +52,11 @@
     }
 
     public static FileType newFileType(String... extensions) {
+        for (int i = 0; i < extensions.length; i++) {
```
A normal for each loop should also work

NorwayMaple replied:
Got it.

NorwayMaple replied:
From what I've read online, you can't modify elements of an array in-place with a foreach loop, which is what I'm doing. See the line `extensions[i] = extensions[i].substring(extensions[i].indexOf('.') + 1);`. Am I mistaken?

Siedlerchr commented:
Pleas ressolve the l10n errors/conflicts

tobiasdiez commented on src/main/java/org/jabref/gui/exporter/ExportCustomizationDialogView.java:
```diff
@@ -0,0 +1,82 @@
+        extensionColumn.setCellValueFactory(cellData -> cellData.getValue().getExtension());
+    }
+
+    private void closeDialog() {
```
Inline this method?

tobiasdiez commented on src/main/java/org/jabref/gui/exporter/ExportCustomizationDialogViewModel.java:
```diff
@@ -0,0 +1,98 @@
+        init();
+    }
+
+    public void loadExporters() {
```
I would prefer if this method can be made private.

tobiasdiez commented on src/main/java/org/jabref/gui/exporter/CreateModifyExporterDialogView.java:
```diff
@@ -0,0 +1,73 @@
+
+import com.airhacks.afterburner.views.ViewLoader;
+
+public class CreateModifyExporterDialogView extends BaseDialog<Optional<ExporterViewModel>> {
```
This should be `BaseDialog<ExporterViewModel>` as the `show` method already returns an `Optional` (which is empty if the user choose to close the dialog using e.g. the cancel button). Then, return `null` in `setResultConverter` when the user canceled the action or the exporter is invalid.

NorwayMaple replied:
I just pushed a fix for this.

tobiasdiez commented on src/main/java/org/jabref/gui/exporter/ExportCustomizationDialogViewModel.java:
```diff
@@ -0,0 +1,98 @@
+
+    public void modifyExporter() {
+        CreateModifyExporterDialogView dialog;
+        ExporterViewModel exporterToModify = selectedExporters.get(0);
```
This throws an exception when no exporter is selected. I start the method with `if (selectedExporters.isEmpty()) {return;}` because there is nothing that we can do if the users does not tell us what exporter to edit.

tobiasdiez commented on src/main/java/org/jabref/gui/exporter/ExportCustomizationDialogViewModel.java:
```diff
@@ -0,0 +1,98 @@
+
+    public void saveToPrefs() {
+        List<TemplateExporter> exportersLogic = new ArrayList<>();
+        exporters.forEach(exporter -> exportersLogic.add(exporter.getLogic()));
```
I find `List<TemplateExporter> exportersLogic = exporters.stream().map(ExporterViewModel:getLogic).collect(toList());` more readable, but this is a matter of taste.

tobiasdiez commented on src/main/java/org/jabref/gui/exporter/ExportCustomizationDialogViewModel.java:
```diff
@@ -0,0 +1,98 @@
+        return exporters;
+    }
+
+    public void init() {
```
inline?

tobiasdiez commented on src/main/java/org/jabref/gui/exporter/ExporterViewModel.java:
```diff
@@ -0,0 +1,49 @@
+
+    private final TemplateExporter exporter;
+    private final StringProperty name = new SimpleStringProperty();
+    private final StringProperty lfFileName = new SimpleStringProperty();
```
Rename to `layoutFileName`?

tobiasdiez commented on src/main/java/org/jabref/gui/exporter/ExporterViewModel.java:
```diff
@@ -0,0 +1,49 @@
+        return this.exporter;
+    }
+
+    public StringProperty getName() {
```
The convention is that methods returning the property should end with `*Property` and do not start with `get`, i.e `nameProperty()` here.

tobiasdiez commented on src/main/java/org/jabref/logic/exporter/TemplateExporter.java:
```diff
@@ -81,7 +97,11 @@ public TemplateExporter(String displayName, String consoleName, String lfFileNam
             LayoutFormatterPreferences layoutPreferences, SavePreferences savePreferences) {
         super(consoleName, displayName, extension);
-        this.lfFileName = Objects.requireNonNull(lfFileName);
+        if (Objects.requireNonNull(lfFileName).endsWith(".layout")) {
```
Maybe extract `".layout"` to a static constant?

Siedlerchr replied:
Eclipse helps you with refactoring. Choose refactor, extract to constant and then you have the option to replace all occurrences with the constant

NorwayMaple replied:
I added a few constants, since `".layout"` was not the online constant string used.

tobiasdiez commented on src/main/java/org/jabref/preferences/JabRefPreferences.java:
```diff
@@ -1983,4 +1986,67 @@ public FileDragDropPreferenceType getEntryEditorFileLinkPreference() {
+            exporterName = formatData.get(EXPORTER_NAME_INDEX);
+            filename = formatData.get(EXPORTER_FILENAME_INDEX);
+            extension = formatData.get(EXPORTER_EXTENSION_INDEX);
+            Optional<TemplateExporter> format = Optional.of(new TemplateExporter(exporterName, filename, extension,
```
This is somewhat strange. If you use `Optional.of(new ...)` then the optional is always non-empty. Thus, there is no need to use an optional to start with.

NorwayMaple commented:
Thank you both for your comments!  I will look into them soon.  @Siedlerchr, would you still like to Skype or do you think you have looked at the code enough?  I am glad to proceed either way.

NorwayMaple commented on src/main/java/org/jabref/gui/exporter/CreateModifyExporterDialog.fxml:
```diff
@@ -0,0 +1,54 @@
+
+<DialogPane maxHeight="-Infinity" maxWidth="-Infinity" minHeight="-Infinity" minWidth="-Infinity" prefHeight="300.0" prefWidth="480.0" xmlns="http://javafx.com/javafx/8.0.171" xmlns:fx="http://javafx.com/fxml/1" fx:controller="org.jabref.gui.exporter.CreateModifyExporterDialogView">
+    <header>
+      <AnchorPane minHeight="0.0" minWidth="0.0" />
```
None. Scene Builder created it automatically and I didn't remove it. I have now removed it.

NorwayMaple commented on src/main/java/org/jabref/gui/exporter/CreateModifyExporterDialog.fxml:
```diff
@@ -0,0 +1,54 @@
+
+<DialogPane maxHeight="-Infinity" maxWidth="-Infinity" minHeight="-Infinity" minWidth="-Infinity" prefHeight="300.0" prefWidth="480.0" xmlns="http://javafx.com/javafx/8.0.171" xmlns:fx="http://javafx.com/fxml/1" fx:controller="org.jabref.gui.exporter.CreateModifyExporterDialogView">
+    <header>
+      <AnchorPane minHeight="0.0" minWidth="0.0" />
```
See above.

NorwayMaple commented on src/main/java/org/jabref/gui/exporter/CreateModifyExporterDialog.fxml:
```diff
@@ -0,0 +1,54 @@
+                  <Label text="Main Layout File:" GridPane.rowIndex="2" />
+                  <TextField fx:id="fileName" GridPane.rowIndex="3" />
+                  <TextField fx:id="extension" GridPane.rowIndex="6" />
+                  <Label alignment="TOP_LEFT" text="Export format name:">
```
Done

NorwayMaple commented on src/main/java/org/jabref/gui/exporter/CreateModifyExporterDialogView.java:
```diff
@@ -0,0 +1,73 @@
+    private final JournalAbbreviationLoader loader;
+
+    public CreateModifyExporterDialogView(Optional<ExporterViewModel> exporter, DialogService dialogService,
+                                          PreferencesService preferences, JournalAbbreviationLoader loader) { //should the latter three have been injected as in the main dialog rather than passed as a param?
```
Done. I have also injected `dialogService` and `preferences`.

NorwayMaple commented on src/main/java/org/jabref/gui/exporter/CreateModifyExporterDialogView.java:
```diff
@@ -0,0 +1,73 @@
+            }
+        });
+
+        browseButton.setOnAction(event -> browse());
```
Done.

NorwayMaple commented on src/main/java/org/jabref/gui/exporter/CreateModifyExporterDialog.fxml:
```diff
@@ -0,0 +1,54 @@
+            </GridPane>
+         </children></AnchorPane>
+    </content>
+    <ButtonType fx:id="saveExporter" text="%Save exporter" />
```
Done.

NorwayMaple replied:
This hasn't worked despite my adding`<ButtonType fx:constant="CANCEL" />` to the FXML, and I'm not sure why. The button does not appear in the dialog.

NorwayMaple commented on src/main/java/org/jabref/gui/exporter/CreateModifyExporterDialogView.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @FXML
+    private void closeDialog() {
```
I thought the FXML used it. I can try taking it out...

NorwayMaple commented on src/main/java/org/jabref/gui/exporter/CreateModifyExporterDialogViewModel.java:
```diff
@@ -0,0 +1,115 @@
+    private final JournalAbbreviationLoader loader;
+
+
+    public CreateModifyExporterDialogViewModel(Optional<ExporterViewModel> exporter, DialogService dialogService, PreferencesService preferences,
```
I understand. There's a bigger question about whether I have used `Optional<T>` correctly, since I have used it to combine two functionalities into one class. If I use omit the argument, I may have to either initialize `exporter` to `null` in the variable declarations or write a number of duplicate methods for each situation. I can talk about this with Christoph on Skype as well.

tobiasdiez replied:
It's perfectly valid to unify two use cases in one class and provide distinct functionality. In this case, you can make the *variable* `exporter` an optional which may or may not be present. My remark concerned the usage of `Optional` as a method *argument*.

NorwayMaple replied:
I just pushed this.

NorwayMaple commented on src/main/java/org/jabref/gui/exporter/CreateModifyExporterDialogViewModel.java:
```diff
@@ -0,0 +1,115 @@
+            .addExtensionFilter(Localization.lang("Custom layout file"), StandardFileType.LAYOUT)
+            .withDefaultExtension(Localization.lang("Custom layout file"), StandardFileType.LAYOUT)
+            .withInitialDirectory(getExportWorkingDirectory()).build();
+        dialogService.showFileOpenDialog(fileDialogConfiguration).ifPresent(f -> layoutFile.set(f.toAbsolutePath().toString())); //implement setting the text
```
Right now it's called in as part of `fileDialogConfiguration` in the line above. How would I make that change?

NorwayMaple replied:
See above.

NorwayMaple commented on src/main/java/org/jabref/gui/exporter/CreateModifyExporterDialogViewModel.java:
```diff
@@ -0,0 +1,115 @@
+        dialogService.showFileOpenDialog(fileDialogConfiguration).ifPresent(f -> layoutFile.set(f.toAbsolutePath().toString())); //implement setting the text
+    }
+
+    private void setTextFields(Optional<ExporterViewModel> exporter) {
```
Do you mean the method `browse`? What does "inline" mean?

tobiasdiez replied:
https://refactoring.com/catalog/inlineFunction.html

NorwayMaple commented on src/main/java/org/jabref/gui/exporter/CreateModifyExporterDialogViewModel.java:
```diff
@@ -0,0 +1,115 @@
+            .addExtensionFilter(Localization.lang("Custom layout file"), StandardFileType.LAYOUT)
+            .withDefaultExtension(Localization.lang("Custom layout file"), StandardFileType.LAYOUT)
+            .withInitialDirectory(getExportWorkingDirectory()).build();
+        dialogService.showFileOpenDialog(fileDialogConfiguration).ifPresent(f -> layoutFile.set(f.toAbsolutePath().toString())); //implement setting the text
```
I put it in both, since saving the exporter might also indicate where the working directory should be. However, now it won't set the directory as before. Since this fix isn't working I am pushing this into a separate branch `exporter_to_javafx_setwd` (in my fork, but not in this pull request). I will try to get to the other changes you and Christoph mentioned later.

NorwayMaple commented on src/main/java/org/jabref/gui/exporter/ExportCustomizationDialog.fxml:
```diff
@@ -0,0 +1,29 @@
+         </children>
+      </VBox>
+   </content>
+   <ButtonType fx:id="addButton" text="%Add" />
```
I made some changes (saved to branch in my fork "exporter_to_javafx_move_buttons"), and the dialog appears with my changes in Scene Builder, but it looks the same when I open it from Eclipse. I'm not sure what's going on, but I will continue with the other changes you and Christoph recommended.

NorwayMaple commented:
I made a number of the changes. For some reason, the cancel button is not showing up in the subdialog. Also, I tried to use a try/catch block to catch an `IndexOutOfBoundsException` when `selectedExporters.get(0)` is called on an empty list, i.e. no exporters are selected. However the exception doesn't get caught. Separately (but not pushed), I tried just doing a check for it `if (selectedExporters.isEmpty()) {return;}`, but that led to the same uncaught exception.

NorwayMaple commented on src/main/java/org/jabref/gui/exporter/ExportCustomizationDialogViewModel.java:
```diff
@@ -0,0 +1,91 @@
+        try {
+            exporterToModify = selectedExporters.get(0);
+            dialog = new CreateModifyExporterDialogView(exporterToModify, dialogService, preferences, loader);
+        } catch (IndexOutOfBoundsException ex) {
```
It's not clear. I put debug breakpoints on lines 56, 57, 59, and 60. I also put one in the line `viewModel.modifyExporter()` in `ExporterCustomizationDialogView`.  That one is the only one it stops at before it throws the exception, but I don't understand how that is possible, since there should be no intervening lines. Also I find it odd that it should throw the very same exception that I was trying to catch.

tobiasdiez replied:
You can try to make `selectedExporters` a normal array list.

NorwayMaple replied:
I don't think it has to do with `selectedExporters` because it returns at line 59 when none are selected. Something else may be going on.

NorwayMaple commented:
@Siedlerchr @tobiasdiez I think I have finished the corrections you guys recommended. Do you think anything remains to be changed?

Thanks,
Abe

tobiasdiez commented on src/main/java/org/jabref/gui/exporter/ExportCustomizationDialogView.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Inject private DialogService dialogService;
+    @Inject private PreferencesService preferences;
+    @Inject private JournalAbbreviationLoader loader; // Should this be injected?
```
e.g here

tobiasdiez commented on src/main/java/org/jabref/gui/exporter/ExportCustomizationDialogViewModel.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    private void loadExporters() {
+        List<TemplateExporter> exportersLogic = preferences.getCustomExportFormats(loader); //Var exportersLogic may need more descriptive name
```
or here

NorwayMaple commented:
I removed most of the comments - let me know if I need to remove the last few. For the code formatter, I ran it, then `git reset --hard` and just removed some extra newlines (it wanted me to add indents that didn't make sense as well). Thanks for your help!
